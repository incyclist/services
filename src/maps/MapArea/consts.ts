
export const RADIUS_EARTH = 6378.1;
export const DEFAULT_RADIUS = 1000; // 1000m default radius
export const DEFAULT_MIN_WAYS = 70;
export const DEFAULT_MAX_WAYS = 300;
export const MAX_DISTANCE_FROM_PATH = 5; // 5m is considered to be max distance a point can be off a path to be counted "inside"
export const GET_WAYS_IN_AREA = '[out:json];way[highway](__boundary__);(._;>;);out geom;';
export const DEFAULT_FILTER = [ 'construction','footway','service','pedestrian', 'path', 'cycleway','elevator','steps','track','escape','bus_guideway','busway','bridleway','corridor','via_ferrata'];
export const DEFAULT_POSITION ={ lat:26.68537966608196,lng:-80.03481961660725 } //Palm  Beach

