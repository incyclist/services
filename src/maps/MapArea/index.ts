export * from './service'
export * from './MapArea'