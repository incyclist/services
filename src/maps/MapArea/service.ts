import { OverpassApi } from "../../api";
import { distanceBetween } from "../../utils/geo";
import { Observer, Singleton } from "../../base/types";
import { IncyclistService } from "../../base/service";
import { FreeRideDataSet,  IncyclistNode, FreeRidePosition, IMapAreaService, IMapArea } from "./types";
import { DEFAULT_MIN_WAYS, DEFAULT_MAX_WAYS, DEFAULT_RADIUS, GET_WAYS_IN_AREA, MAX_DISTANCE_FROM_PATH, DEFAULT_FILTER} from "./consts";
import { buildQuery,getBounds, parseMapData } from "./utils";
import { Injectable } from "../../base/decorators";
import { MapArea } from "./MapArea";
import { OptionManager } from "./options";


type MapAreaRecord = {
    map: MapArea
    lastUsed: number
    radius: number
}
const MAX_MAPS = 5

@Singleton
export class MapAreaService extends IncyclistService implements IMapAreaService {
    protected static consts = { DEFAULT_RADIUS,DEFAULT_MIN_WAYS,DEFAULT_MAX_WAYS,MAX_DISTANCE_FROM_PATH,GET_WAYS_IN_AREA }


    protected overpass:OverpassApi
    protected loaded: 'success' | 'failure' |'unknown' = 'unknown'
    protected observer: Observer
    protected current: MapArea
    
    
    protected minWays: number
    protected maxWays: number
    protected radius: number
    protected filter:Array<string>
    protected maps: Record<string,MapAreaRecord> 
    protected optionManager: OptionManager
    protected iv: NodeJS.Timeout


    constructor () {
        super('MapArea')

        this.minWays = DEFAULT_MIN_WAYS
        this.maxWays = DEFAULT_MAX_WAYS
        this.radius  = DEFAULT_RADIUS
        this.filter  = DEFAULT_FILTER
        this.maps = {}
    }

    /**
     * Load a new map area for a given location. If the same location is already loaded, return the existing map.
     * @param location - the location to load the map for
     * @returns a promise that resolves when the map is loaded, or rejects if the load fails
     */
    async load( location: IncyclistNode):Promise<IMapArea> {

        const reloadRequires = this.requiresReload(location)
        if (!reloadRequires) {
            this.getMapForLocation(location).lastUsed = Date.now()
            return this.getMap(location)
        }
        

        let ts
        try {                       
            const bounds = getBounds(location?.lat,location?.lng,this.radius)
            const query = buildQuery(GET_WAYS_IN_AREA,bounds);
            const {id,lat,lng} = location;


            this.logEvent({message:'overpass query',query,location:{id,lat,lng}, radius:this.radius});
            console.log('load map',location,query, this.radius)
            ts = Date.now();
            const openmapData = await this.getOverpassAPI().query(query);                  
            this.logEvent({message:'overpass query result',status:'success',duration:(Date.now()-ts)});

            const data = this.createMapData(openmapData);    
            const map = new MapArea(data,location,bounds)
            this.addMap(map,location)
            this.getOptionManager().setMap(map)

            this.checkRadiusAdjustment(map);
            
            this.startGarbageCollection()

            return map
        }
        catch (error) {
            const duration = ts? Date.now()-ts : undefined
            this.logEvent({message:'overpass query result',status:'failure',error:{code:error.code,response:error.response},duration});
        }

    }

    getMap(location:IncyclistNode):MapArea {
        const maps = this.getAllMaps()??[]
        return maps.find(m=>m.isWithinBoundary(location))        
    }

    getOptionManager():OptionManager {
        this.optionManager = this.optionManager ?? new OptionManager(this)
        return this.optionManager
    }

    setFilter(filter:Array<string>|null) {
        this.filter = filter??undefined
    }


    protected requiresReload(location:IncyclistNode):boolean {
        if (!this.hasLocationCovered(location))
            return true

        const record = this.getMapForLocation(location)
        const {map,radius} = record
        if (!map)
            return true

        const dist = distanceBetween(location,map.getQueryLocation())
        const updateRequired = dist>(radius/5); // 20%
        this.logEvent( {message:'distance between previous overpass request',dist,updateRequired, pct:dist/radius*100});

        return updateRequired
    }

   

    protected hasLocationCovered(location:IncyclistNode):boolean {
        const maps = this.getAllMaps()??[]
        return maps.some(m=>m.isWithinBoundary(location))        
    }


    protected mapsKey(location:IncyclistNode):string {
        return `${location.lat},${location.lng}`        
    }

    protected getAllMaps():MapArea[] {
        return Object.values(this.maps).map(m=>m.map)
    }

    protected addMap(map:MapArea,location:IncyclistNode) {

        this.current = map

        // replace existing map if it exisst
        const existing = Object.values(this.maps).findIndex(m=>m.map.isWithinBoundary(location))        
        if (existing!==-1) {
            const key  = Object.keys(this.maps)[existing]
            delete this.maps[key]
        }

        this.maps[this.mapsKey(location)] = {map,lastUsed:Date.now(), radius:this.radius}
    }

    protected getMapForLocation(location:IncyclistNode):MapAreaRecord {        
        return Object.values(this.maps).find(m=>m.map.isWithinBoundary(location))        
    }



    createMapData( openmapData):FreeRideDataSet {
        

        let ts = Date.now();
        const data = parseMapData(openmapData,this.filter); 
        let ts1 = Date.now(); 

        /* istanbul ignore next */ 
        this.logger.logEvent({message:'Parse',duration:(ts1-ts),
                                ways:data?.ways.length??0,
                                nodes:Object.keys(data?.nodesLookup??{}).length,
                                typeStats:data?.typeStats
                            });

        if (data!==undefined) {
            this.loaded = 'success';
        }
        else {
            this.loaded = 'failure'
        }

        return data

    }

    /**
     * Check if the number of ways found in the map is within the expected range (minWays,maxWays).
     * If the number of ways is below the minimum, adjust the radius to find more ways during next query
     * If the number of ways is above the maximum, adjust the radius to find fewer ways  during next query
     * @param map The map to check
     */
    protected checkRadiusAdjustment(map:MapArea) {
        const ways = map.getWays()
        if (ways.length < this.minWays) {
            if (ways.length > 0) {
                let gap = this.minWays / ways.length;
                this.radius = this.radius * Math.sqrt(gap);
            }
            else {
                this.radius = this.radius * 2;
            }
        }
        else if (ways.length > this.maxWays) {
            let gap = ways.length / this.maxWays;
            this.radius = this.radius / Math.sqrt(gap);
        }
    }

    protected startGarbageCollection() {
        if (this.iv)
            return;

        this.iv = setInterval(() => {
            this.garbageCollection()
        }, 1000 * 60 * 5);
    }

    protected stopGarbageCollection() {
        if (!this.iv)
            return;
        
        clearInterval(this.iv)
        this.iv = undefined 
    }

    protected garbageCollection() {
        const now = Date.now();
        const maps = Object.values(this.maps).filter(m=>now-m.lastUsed<1000*60*5);

        // Always keep the last map (this.current)  even if it has not been used for a while
        const keyLastUsed = this.mapsKey(this.current.getQueryLocation())        
        const deleteTarget  = maps.filter(m=>this.mapsKey(m.map.getQueryLocation())!==keyLastUsed)

        if (deleteTarget.length>MAX_MAPS)
            deleteTarget.forEach(m=>delete this.maps[this.mapsKey(m.map.getQueryLocation())])        
    }



    @Injectable
    protected getOverpassAPI() {
        this.overpass = this.overpass??new OverpassApi();
        return this.overpass

    }


    reset() {
        this.stopGarbageCollection()    
        super.reset()

    }

}



export const useMapArea = () => new MapAreaService()