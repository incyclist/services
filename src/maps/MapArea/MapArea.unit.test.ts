import fs from 'fs/promises';

import {useMapArea} from './service';
import testData from '../../../__tests__/data/overpass/MapArea-test1.json';
import defaultData from '../../../__tests__/data/overpass/default-location.json';
import roundaboutData from '../../../__tests__/data/overpass/roundabout-test.json';

import { MapArea } from './MapArea';
import {  IncyclistNode, IncyclistWay, IncyclistWaySplit, PathCrossingInfo, WayInfo } from './types';
import { getPointCrossingPath, isOneWay } from './utils';


const printWay = (way:WayInfo) => {
    const info = way.id+'='+way.path.map((m) => `${m.id??'crossing'}`).join(',')
    return info
}

describe( 'MapArea', () => {
    describe ( 'constructor and getters', () => {

        let area:MapArea
        beforeEach(() => {           
            useMapArea().setFilter(null)
        });

        afterEach(() => {
            useMapArea().reset()
            jest.resetAllMocks()
        });

        test('normal flow', () => { 
            const location = { lat:1, lng:1 }
            const boundary = { northeast:{ lat:1, lng:1 }, southwest:{ lat:1, lng:1 } }
            const data = useMapArea().createMapData(testData);           
            area = new MapArea(data, location, boundary);            

            expect(area.getQueryLocation()).toEqual(location);
            expect(area.getBoundary()).toEqual(boundary);
            expect(area.getStats()).toEqual({"footway": 1, "path": 1, "service": 1, "unclassified": 1});

            expect(area.getWays().map(w => w.id)).toMatchSnapshot()
            expect(area.getWay('4815077')).toMatchSnapshot()
            expect(area.getNode('30926693')).toMatchSnapshot()
            expect(area.getWay('non-existing')).toBeUndefined()
            expect(area.getNode('non-existing')).toBeUndefined()

        })

        test('with multi-way roundabouts', () => { 
            const location = { lat:47.73444191318148,lng:1.6481140252482982 }
            const boundary = { northeast:{ lat:1, lng:1 }, southwest:{ lat:1, lng:1 } } // not relevant
            const data = useMapArea().createMapData(roundaboutData);           
            area = new MapArea(data, location, boundary);            

            const roundabout = area.getWay('700616380')
            expect(roundabout).toBeUndefined()

            const multiNodeRB = area.getWays().find(w => w.id.startsWith('R:')&&w.id.split(',').includes('700616380')) ??{} as IncyclistWay
            expect(printWay(multiNodeRB)).toMatchSnapshot()
            
            

        })


    })

    describe ( 'is within boundary', () => {

        let area:MapArea
        beforeEach(() => {           
            useMapArea().setFilter(null)
        });

        afterEach(() => {
            useMapArea().reset()
            jest.resetAllMocks()
        });

        test('normal flow', () => { 
            const location = { lat:1, lng:1 }
            const boundary = { northeast:{ lat:20, lng:20 }, southwest:{ lat:0, lng:0 } }
            const data = useMapArea().createMapData(testData);           
            area = new MapArea(data, location, boundary);            

            expect(area.isWithinBoundary({ lat:1, lng:1 })).toBe(true);
            expect(area.isWithinBoundary({ lat:21, lng:1 })).toBe(false);
            expect(area.isWithinBoundary({ lat:-1, lng:2 })).toBe(false);
            expect(area.isWithinBoundary({ lat:10, lng:21 })).toBe(false);
            expect(area.isWithinBoundary({ lat:10, lng:-1 })).toBe(false);

        })


    })

    describe ( 'getNearestPath', () => {
        
        let area: MapArea;
        let data
        let testWays


        
        beforeEach(() => {           
            useMapArea().setFilter(null)
            data = useMapArea().createMapData(testData);
            testWays = data.ways;
            // position and bounds are irrelevant for this test
            area = new MapArea(data, { lat:0, lng:0 }, { northeast:{ lat:0, lng:0 }, southwest:{ lat:0, lng:0 } });            
        });

        afterEach(() => {
            useMapArea().reset()
            jest.resetAllMocks()
        });
    
    
        test ( 'on 1st point in path #1', () => {


            let res = area.getNearestPath( testWays[0].path[0]);
            expect ( res.path ).toBeDefined();
            expect ( res.distance ).toBe(0);
            expect ( res.path[0].lat ).toBe(testWays[0].path[0].lat);
            expect ( res.path[0].lng ).toBe(testWays[0].path[0].lng);
    
    
        } );
        test ( 'on 1st point in path #2', () => {
    
            let res = area.getNearestPath(testWays[1].path[0] );
            expect ( res.path ).toBeDefined();
            expect ( res.distance ).toBe(0);
            expect ( res.path[0].lat ).toBe(testWays[1].path[0].lat);
            expect ( res.path[0].lng ).toBe(testWays[1].path[0].lng);
    
    
        } );
        test ( 'rd point on 1st path', () => {
            let location = { lat: -16.7824429, lng: 145.6983982 };
            let res = area.getNearestPath(location );

            expect ( res.path ).toBeDefined();
            expect ( res.distance ).toBe(0);
            expect ( res.path[0].lat ).toBe(testWays[0].path[0].lat);
            expect ( res.path[0].lng ).toBe(testWays[0].path[0].lng);
            
        } );
    

        test ( 'two ways with same distance -> delivers first match', () => {
            const distanceToPathSpy = jest.spyOn(require('./utils'), 'distanceToPath').mockReturnValue(1);
    
            let location = { lat: -16.7824429, lng: 145.6983982 };
            const res = area.getNearestPath(location);
    
            expect(res.way.id).toBe('4815077');
            expect(res.distance).toBe(1);

            distanceToPathSpy.mockRestore();
        });

        test ( 'error: distanceToPath will have lowest value for way with highest id', () => {
            const distanceToPathSpy = jest.spyOn(require('./utils'), 'distanceToPath').mockImplementation((point,w) => { return 184561986-Number((w as IncyclistWay).id); });
    
            
            let location = { lat: -16.7824429, lng: 145.6983982 };
            const res = area.getNearestPath(location);
    
            expect(res.way.id).toBe('184561984');
            expect(res.distance).toBe(2)
            distanceToPathSpy.mockRestore();
    
        });

    
        test ( 'error: distanceToPath returns undefined', () => {
            const distanceToPathSpy = jest.spyOn(require('./utils'), 'distanceToPath').mockReturnValue(undefined);
    
            let location = { lat: -16.7824429, lng: 145.6983982 };
            const res = area.getNearestPath(location);
    
            expect(res.way).toBeUndefined()
            expect(res.distance).toBeUndefined();
            distanceToPathSpy.mockRestore();
    
        });


    });

    
    describe ( 'splitAtFirstBranch', () => {
        
        let area: MapArea;
        
        beforeEach(() => {           
            useMapArea().setFilter(null)
            const data = useMapArea().createMapData(defaultData);
            // position and bounds are irrelevant for this test
            const location = {lat: 26.685379666082877,lng: -80.03481961660714}
            const boundary = { northeast:{ lat:0, lng:0 }, southwest:{ lat:0, lng:0 } } // does not matter for this test
            area = new MapArea(data, location,boundary );            
        });

        afterEach(() => {
            useMapArea().reset()
            jest.resetAllMocks()
        });

        const addWay = (area,way) => {
            
            area.data.waysLookup[way.id] = way
            area.data.ways.push(way)
            for (const n of way.path) {
                if (!area.data.nodesLookup[n.id]) {
                    area.data.nodesLookup[n.id] = n                    
                }
            }
        }
    
    
        test.skip ( 'test to explore the data', () => {
            const node = area.getNode('7539895026')
            const ways = node?.ways??[]
            console.log(ways.map(m => `${area.getWay(m).id}:${area.getWay(m).name}`))

            const way = area.getWay('82696507')
            console.log(way.path.map((m,idx) => `${idx}:${m.id}:${m.lat},${m.lng},${m.ways?.length}`))
    
        } );

        test(('default location'),()=>{
            const way = area.getWay('82696507')
            const result = area.splitAtFirstBranch(way);

            expect(result.wayId).toBe('82696507')
            expect(result.path).toHaveLength(19)
            
            const info = result.branches?.map((b) => `${b.id}:`+b.path.map((m) => `${m.id}`).join(','))
            expect(info).toMatchSnapshot()
        })

        test(('street crossing a one way street'),()=>{
            const way = area.getWay('82696507')
            const node = way.path[2]

            // construct a one way street crossing the way (crossing point with '82696507' is last point of that way)
            // we are not allowed to drive into that one way street
            const oneWay:IncyclistWay = {
                id:'1234', name:'Dummy Ave',type:'way', 
                tags: {oneway:'yes', highway: 'residential'  },
                bounds:{minlat:0, maxlat:0, minlon:0, maxlon:0 },
                path:[
                { id: 'Dummy1', lat: -16.7824429, lng: 145.6983982, ways:['1234'] },
                { id: 'Dummy2', lat: -16.7824430, lng: 145.6983982, ways:['1234'] },
                node
            ]}
            node.ways?.push(oneWay.id)
            addWay(area,oneWay)
            
            const result = area.splitAtFirstBranch(way);
            expect(result.wayId).toBe('82696507')
            expect(result.path).toHaveLength(19)        // one way should be ignored, as we cannot drive into it            
        })


    });    


    describe ( 'splitAtCrossingPoint', () => {
        
        let area: MapArea;
        let data,location,boundary  
        
        beforeEach(() => {           
            // position and bounds are irrelevant for this test
            location = {lat: 26.685379666082877,lng: -80.03481961660714}
            boundary = { northeast:{ lat:0, lng:0 }, southwest:{ lat:0, lng:0 } } // does not matter for this test
        });

        afterEach(() => {
            useMapArea().reset()
            jest.resetAllMocks()
        });
    
        test.skip ( 'test to explore the data', () => {

            console.log('-------------------- TEST CASE SHOULD NOT BE PART OF A SUITE RUN')            
            expect(process.env.CI).toBeUndefined()  
            

            data = useMapArea().createMapData(defaultData);
            area = new MapArea(data, location,boundary );            

            const node = area.getNode('7539895026')
            const ways = node?.ways??[]
            console.log(ways.map(m => `${area.getWay(m).id}:${area.getWay(m).name}`))

            const way = area.getWay('82696507')
            console.log(way.path.map((m,idx) => `${idx}:${m.id}:${m.lat},${m.lng},${m.ways?.length}`))
    
        } );

    

        test(('default location'),()=>{
            data = useMapArea().createMapData(defaultData);
            area = new MapArea(data, location,boundary );            

            const way = area.getWay('82696507')
            const first = way.path[0]
            const last = way.path[way.path.length-1]

            const crossing = getPointCrossingPath(location, way.path,true)
            const result = area.splitAtCrossingPoint(way, crossing);

            

            // Expected: 2 results
            // 1st: crossing point -> first point of way (in reverse order)
            // 2nd: crossing point -> last  point of way
            expect(result[0].wayId).toBe('82696507')
            expect(result[0].path?.[0]).toMatchObject(crossing.point)
            expect(result[0].path?.[result[0].path?.length-1]?.id).toEqual(first.id)

            expect(result[1].wayId).toBe('82696507')
            expect(result[0].path?.[0]).toMatchObject(crossing.point)
            expect(result[1].path?.[result[1].path?.length-1]?.id).toEqual(last.id)            
        })


        test(('crossing is close to last point'),()=>{
            data = useMapArea().createMapData(defaultData);
            area = new MapArea(data, location,boundary );            

            const way = area.getWay('82696507')
            const first = way.path[0]
            const last = way.path[way.path.length-1]    // see https://www.openstreetmap.org/node/99649684

            const crossing = getPointCrossingPath(last, way.path,true)  
            crossing.distance = 10
            const result = area.splitAtCrossingPoint(way, crossing);

            // Expected: 3 results (as there are two streets crossing at the last point of the way)
            
            // 1st: crossing point -> first point of way (in reverse order)
            // 2nd: way crossing the street at last point
            // 2nd: way crossing the street at last point
            expect(result[0].wayId).toBe('82696507')
            expect(result[0].path?.[0]).toMatchObject(crossing.point)
            expect(result[0].path?.[result[0].path?.length-1]?.id).toEqual(first.id)

            expect(result[1].wayId).toBe('11205160')
            expect(result[2].wayId).toBe('47346032')
        })

        test(('crossing is exactly at first point, street crosses another street in the middle'),()=>{
            data = useMapArea().createMapData(defaultData);
            area = new MapArea(data, location,boundary );            

            const way = area.getWay('11236601')
            const first = way.path[0]    // see https://www.openstreetmap.org/node/99836339
            const last = way.path[way.path.length-1]    

            const crossing = getPointCrossingPath(first, way.path,true)  
            const result = area.splitAtCrossingPoint(way, crossing);

            // Expected: 3 results (as there are is a streets crossing at the first point of the way)
            
            // 1st: way crossing the street at first point, from crossing->first (reversed)
            // 2nd: way crossing the street at first point, from crossing->last 
            // 3rd: full way 11236601
            expect(result[0].wayId).toBe('82696507')
            expect(result[1].wayId).toBe('82696507')
            
            expect(result[2].wayId).toBe('11236601')
            expect(result[2].path?.length).toEqual(way.path.length)
            expect(result[2].path?.[0]?.id).toEqual(first.id)
            expect(result[2].path?.[result[2].path?.length-1]?.id).toEqual(last.id)


        })

        test('crossing is exactly a point in the path of the one way street',()=>{ 
            
            useMapArea().setFilter(null)
            data = useMapArea().createMapData(defaultData);
            area = new MapArea(data, location,boundary );            
            
            const way = area.getWay('907331966')
            const {lat,lng,id:nodeId} = way.path[3]
            const last = way.path[way.path.length-1]

            const crossing = getPointCrossingPath({lat,lng}, way.path,true)  
            const result = area.splitAtCrossingPoint(way, crossing);
            
            expect(result).toHaveLength(1)
            expect(result[0].wayId).toBe('907331966')
            expect(result[0].path?.[0].id).toEqual(nodeId)
            expect(result[0].path?.[result[0].path?.length-1]?.id).toEqual(last.id)

        })



        test(('one way'),()=>{
            useMapArea().setFilter(null)    // if we apply filters, we would not find a oneway in this area
            data = useMapArea().createMapData(defaultData);
            area = new MapArea(data, location,boundary );            

            const way = area.getWay('907331966')    // this is the new way route
            const last = way.path[way.path.length-1]
            const startPos = {lat:26.69212640839961,lng:-80.03713846327933}

            

            const crossing = getPointCrossingPath(startPos, way.path,true)
            const result = area.splitAtCrossingPoint(way, crossing);

            // Expected: 1 result
            // crossing point -> last  point of way
            expect(result).toHaveLength(1)

            expect(result[0].wayId).toBe('907331966')
            expect(result[0].path?.[0]).toMatchObject(crossing.point)
            expect(result[0].path?.[result[0].path?.length-1]?.id).toEqual(last.id)

        })


        test('implicit roundabout',()=>{
            useMapArea().setFilter(null)    // if we apply filters, we would not find a roundabout in this area
            useMapArea().setFilter(null)
            data = useMapArea().createMapData(defaultData);
            area = new MapArea(data, location,boundary );            

            const way = area.getWay('1328903863')
            const startPos = {lat:26.676754683691712,lng:-80.03758371159269}
            const crossing = getPointCrossingPath(startPos, way.path,true)
            const first = way.path[0]
           
            const result = area.splitAtCrossingPoint(way, crossing);

            // Expected: 2 results
            // 1st: crossing point -> first point of way (in reverse order)
            // 2nd: crossing point -> first  point of way( = last point of way)
            expect(result[0].wayId).toBe('1328903863')
            expect(result[0].path?.[0]).toMatchObject(crossing.point)
            expect(result[0].path?.[result[0].path?.length-1]?.id).toEqual(first.id)

            expect(result[1].wayId).toBe('1328903863')
            expect(result[0].path?.[0]).toMatchObject(crossing.point)
            expect(result[1].path?.[result[1].path?.length-1]?.id).toEqual(first.id)            
        })


        // {lat:51.18043270952056,lng:6.40791009405025}

        

    });    

    describe ( 'buidlSegmentInfo', () => {
        
        let area: MapArea;
        let data,location,boundary  

        const prepareOptions = (wayId, crossingPoint, before,after) => {
            const result:IncyclistWaySplit[] = []
            let path:Array<IncyclistNode> = []
            if (before) {
                path.push (...before.map(n => area.getNode(n.toString())))                
                path.push(crossingPoint)
                result.push({wayId,path})
            }
            if (after) {
                path = [ crossingPoint  ]
                path.push (...after.map(n => area.getNode(n.toString())))                
                result.push({wayId,path})
                    
            }
            return result
            
        }

        
        beforeEach(() => {           
            // position and bounds are irrelevant for this test
            location = {lat: 26.685379666082877,lng: -80.03481961660714}
            boundary = { northeast:{ lat:0, lng:0 }, southwest:{ lat:0, lng:0 } } // does not matter for this test
        });

        afterEach(() => {
            useMapArea().reset()
            jest.resetAllMocks()
        });
   

        test(('default location'),()=>{
            data = useMapArea().createMapData(defaultData);
            area = new MapArea(data, location,boundary );            

            const way = area.getWay('82696507')       // see https://www.openstreetmap.org/way/82696507     
            const startPos = {lat:26.685379666083794,lng: -80.03481961660702}            
            const before = [7539895026,8498961197,8498961210,99836339,99887561,99887560,99746466,99887559,99887557,99887555,99887552,99887550,99887548,99887546,99887544,6329797112,99887542,99887540,99887538,99887536,99887534,99887532,99887530,99855503,99735224,99713117,99716395,99750605,99656765,99887528,5077945136,5494360086,99887526,11240951476,5077905518,99751778,5077905520,99755546,99709224,5077945148,99887521,99656394,99628306,5499384959,12375134635]
            const after = [99887565,99887567,99887569,99649684]
            const paths:IncyclistWaySplit[] = prepareOptions(way.id, startPos, before, after)

            const res = area.buildSegmentInfo( way,paths)
            
            expect(res.segments).toHaveLength(2)
            expect(printWay(res.segments[0])).toEqual('82696507=7539895026,8498961197,8498961210,99836339')
            expect(printWay(res.segments[1])).toEqual('82696507=crossing,99887565,99887567')
        })


        test(('one way - no crossing'),()=>{
            useMapArea().setFilter(null)    // if we apply filters, we would not find a oneway in this area
            data = useMapArea().createMapData(defaultData);
            area = new MapArea(data, location,boundary );            

            const way = area.getWay('907331966')    // see https://www.openstreetmap.org/way/82696507   
            const startPos = {lat:26.69212640839961,lng:-80.03713846327933}
            const path = [8425475157,8425475158,8425475159,8425475160,8425475161,8425475162,6329797171]
            const paths:IncyclistWaySplit[] = prepareOptions(way.id, startPos, null, path)
            
            const res = area.buildSegmentInfo( way,paths)
            expect(res.segments).toHaveLength(1)
            expect(printWay(res.segments[0])).toEqual('907331966=crossing,8425475157,8425475158,8425475159,8425475160,8425475161,8425475162,6329797171')
        })



        

    });       
})