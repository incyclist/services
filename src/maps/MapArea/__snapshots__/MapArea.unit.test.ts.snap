// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`MapArea constructor and getters normal flow 1`] = `
[
  "4815077",
  "183973236",
  "183973238",
  "184561984",
]
`;

exports[`MapArea constructor and getters normal flow 2`] = `
{
  "bounds": {
    "maxlat": -16.7820535,
    "maxlon": 145.7023158,
    "minlat": -16.7901496,
    "minlon": 145.6982433,
  },
  "id": "4815077",
  "name": "Vasey Esplanade",
  "path": [
    {
      "id": "1941090209",
      "lat": -16.7820535,
      "lng": 145.6982433,
      "tags": undefined,
      "ways": [
        "4815077",
      ],
    },
    {
      "id": "5701587616",
      "lat": -16.7822514,
      "lng": 145.6983275,
      "tags": undefined,
      "ways": [
        "4815077",
      ],
    },
    {
      "id": "1944102704",
      "lat": -16.7824429,
      "lng": 145.6983982,
      "tags": undefined,
      "ways": [
        "4815077",
        "183973238",
      ],
    },
    {
      "id": "1944102701",
      "lat": -16.7826913,
      "lng": 145.6984974,
      "tags": undefined,
      "ways": [
        "4815077",
        "183973238",
      ],
    },
    {
      "id": "5701587613",
      "lat": -16.7831131,
      "lng": 145.698688,
      "tags": undefined,
      "ways": [
        "4815077",
      ],
    },
    {
      "id": "1941090232",
      "lat": -16.78344,
      "lng": 145.6988323,
      "tags": undefined,
      "ways": [
        "4815077",
      ],
    },
    {
      "id": "5701587612",
      "lat": -16.7834807,
      "lng": 145.6988448,
      "tags": undefined,
      "ways": [
        "4815077",
      ],
    },
    {
      "id": "5701587611",
      "lat": -16.7835291,
      "lng": 145.6988493,
      "tags": undefined,
      "ways": [
        "4815077",
      ],
    },
    {
      "id": "5701587610",
      "lat": -16.7836028,
      "lng": 145.6988474,
      "tags": undefined,
      "ways": [
        "4815077",
      ],
    },
    {
      "id": "5701587609",
      "lat": -16.7837142,
      "lng": 145.6988438,
      "tags": undefined,
      "ways": [
        "4815077",
      ],
    },
    {
      "id": "1941090215",
      "lat": -16.7837544,
      "lng": 145.6988501,
      "tags": undefined,
      "ways": [
        "4815077",
        "184561984",
      ],
    },
    {
      "id": "5701587607",
      "lat": -16.7837984,
      "lng": 145.698862,
      "tags": undefined,
      "ways": [
        "4815077",
      ],
    },
    {
      "id": "30926715",
      "lat": -16.7839079,
      "lng": 145.6989008,
      "tags": undefined,
      "ways": [
        "4815077",
      ],
    },
    {
      "id": "5701587606",
      "lat": -16.7839716,
      "lng": 145.6989271,
      "tags": undefined,
      "ways": [
        "4815077",
      ],
    },
    {
      "id": "5701587605",
      "lat": -16.7840315,
      "lng": 145.6989562,
      "tags": undefined,
      "ways": [
        "4815077",
      ],
    },
    {
      "id": "5701587604",
      "lat": -16.7841513,
      "lng": 145.6990196,
      "tags": undefined,
      "ways": [
        "4815077",
      ],
    },
    {
      "id": "1950581698",
      "lat": -16.7852118,
      "lng": 145.6995816,
      "tags": undefined,
      "ways": [
        "4815077",
        "184561984",
      ],
    },
    {
      "id": "30926693",
      "lat": -16.7852871,
      "lng": 145.6996239,
      "tags": undefined,
      "ways": [
        "4815077",
      ],
    },
    {
      "id": "5701587617",
      "lat": -16.7880341,
      "lng": 145.7010853,
      "tags": undefined,
      "ways": [
        "4815077",
      ],
    },
    {
      "id": "30926716",
      "lat": -16.7882273,
      "lng": 145.7011877,
      "tags": undefined,
      "ways": [
        "4815077",
      ],
    },
    {
      "id": "5701587618",
      "lat": -16.7898716,
      "lng": 145.7022656,
      "tags": undefined,
      "ways": [
        "4815077",
      ],
    },
    {
      "id": "5701587821",
      "lat": -16.7899098,
      "lng": 145.7022967,
      "tags": undefined,
      "ways": [
        "4815077",
      ],
    },
    {
      "id": "30926717",
      "lat": -16.7899532,
      "lng": 145.7023158,
      "tags": undefined,
      "ways": [
        "4815077",
      ],
    },
    {
      "id": "5701587620",
      "lat": -16.7899936,
      "lng": 145.7023088,
      "tags": undefined,
      "ways": [
        "4815077",
      ],
    },
    {
      "id": "5701587822",
      "lat": -16.790026,
      "lng": 145.7022812,
      "tags": undefined,
      "ways": [
        "4815077",
      ],
    },
    {
      "id": "5701587619",
      "lat": -16.7900508,
      "lng": 145.7022536,
      "tags": undefined,
      "ways": [
        "4815077",
      ],
    },
    {
      "id": "30926718",
      "lat": -16.7901496,
      "lng": 145.7021109,
      "tags": undefined,
      "ways": [
        "4815077",
      ],
    },
  ],
  "tags": {
    "highway": "unclassified",
    "name": "Vasey Esplanade",
  },
  "type": "unclassified",
}
`;

exports[`MapArea constructor and getters normal flow 3`] = `
{
  "id": "30926693",
  "lat": -16.7852871,
  "lng": 145.6996239,
  "tags": undefined,
  "ways": [
    "4815077",
  ],
}
`;

exports[`MapArea constructor and getters with multi-way roundabouts 1`] = `"R:24145957,700616375,700616377,700616378,700616379,700616380,950264741=261569371,687166587,3062866649,261569375,687166588,261567886,1667881554,1667881551,3062866659,687166591,261569380,687166592,261569384,687166593,426761660,687166594,261569388,687166595,426761851,687166596,261569371"`;

exports[`MapArea splitAtFirstBranch default location 1`] = `
[
  "82696507:99716395,99713117",
  "11217595:99716395,6329797168",
]
`;
